ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/PullQueryStreamWriter.java

```diff
@@ -0,0 +1,215 @@
+  private boolean closed;
+  private boolean sentAtLeastOneRow = false;
+
+  PullQueryStreamWriter(
+      final PullQueryResult result,
+      final long disconnectCheckInterval,
+      final ObjectMapper objectMapper,
+      final PullQueryQueue pullQueryQueue,
+      final Clock clock,
+      final CompletableFuture<Void> connectionClosedFuture
+  ) {
+    this.result = Objects.requireNonNull(result, "result");
+    this.objectMapper = Objects.requireNonNull(objectMapper, "objectMapper");
+    this.disconnectCheckInterval = disconnectCheckInterval;
+    this.pullQueryQueue = Objects.requireNonNull(pullQueryQueue, "pullQueryQueue");
+    this.clock = Objects.requireNonNull(clock, "clock");
+    connectionClosedFuture.thenAccept(v -> connectionClosed = true);
+    result.onException(t -> pullQueryException = t);
+    result.onCompletion(v -> limitReached = true);
+  }
+
+  @Override
+  public void write(final OutputStream output) {
+    try {
+      StringBuilder sb = new StringBuilder();
+      PullQueryRow head = null;
+      long lastFlush = clock.millis();
+
+      final StreamedRow header
+          = StreamedRow.header(result.getQueryId(), result.getSchema());
+      sb.append("[").append(writeValueAsString(header));
+
+      while (!connectionClosed && !pullQueryQueue.isClosed() && !limitReached) {
+
+        final PullQueryRow row = pullQueryQueue.pollRow(
+            Math.min(disconnectCheckInterval, MAX_FLUSH_MS),
+            TimeUnit.MILLISECONDS
+        );
+        if (row != null) {
+          // We keep track of a head of the queue so that we know when writing the current row if
+          // there a next.  This is to write proper json
+          final PullQueryRow toProcess = head;
+          head = row;
+          if (toProcess == null) {
+            continue;
+          }
+          writeRow(toProcess, head, sb);
+          if (sb.length() >= FLUSH_SIZE_BYTES || (clock.millis() - lastFlush) >= MAX_FLUSH_MS) {
+            output.write(sb.toString().getBytes(StandardCharsets.UTF_8));
+            output.flush();
+            sb = new StringBuilder();
+            lastFlush = clock.millis();
+          }
+        }
+        drainAndThrowOnError(head, output, sb);
+      }
+      if (!pullQueryQueue.isEmpty() || head != null) {
+        drain(head, sb);
+      }
+      sb.append("]");
+      if (sb.length() > 0) {
+        output.write(sb.toString().getBytes(StandardCharsets.UTF_8));
+        output.flush();
+      }
+    } catch (InterruptedException e) {
+      LOG.warn("Interrupted while writing to connection stream");
+    } catch (Throwable e) {
+      LOG.error("Exception occurred while writing to connection stream: ", e);
+      outputException(output, e);
+    }
+  }
+
+  private void writeRow(
+      final PullQueryRow row,
+      final PullQueryRow head,
+      final StringBuilder sb
+  ) {
+    // Send for a comma after the header
+    if (!sentAtLeastOneRow) {
+      sb.append(",").append(System.lineSeparator());
```

Comment:
just wondering, how do we handle content that contains the line separator e.g. a column that has `foo\nbar` as its contents? will that throw anything off? might be worth testing that